.ado/release.yml

```diff
@@ -1,212 +1,269 @@
+              condition: succeeded()
+              targetPath: $(System.DefaultWorkingDirectory)/target/wheels
+              artifactName: wheel.${{ target.os }}.${{ target.arch }}
+          steps:
+          - task: RustInstaller@1
+            inputs:
+              rustVersion: ms-$(RUST_TOOLCHAIN_VERSION)
+              ${{ if target.additionalRustTargets }}:
+                additionalTargets: ${{ target.additionalRustTargets }}
+              cratesIoFeedOverride: $(cratesIoFeedOverride)
+              toolchainFeed: $(toolchainFeed)
+            displayName: Install Rust toolchain
+
+          - script: |
+              rustc --version
+              rustc --print target-list
+            displayName: View rust target info
+            condition: succeeded()
+
+          - task: UsePythonVersion@0
+            inputs:
+              versionSpec: $(PYTHON_VERSION)
+            condition: succeeded()
+
+          # Build or restore LLVM
+          - pwsh: New-Item -ItemType Directory -Path $(LLVM_DIRECTORY) -Force
+            displayName: Ensure cache dir exists
+            condition: succeeded()
+
+          - task: Cache@2
+            inputs:
+              key: '"llvm-$(LLVM_VERSION)-${{ target.os }}-${{ target.arch }}"'
+              restoreKeys: '"llvm-$(LLVM_VERSION)-${{ target.os }}-${{ target.arch }}"'
+              path: $(LLVM_DIRECTORY)
+              cacheHitVar: CACHE_RESTORED
+            displayName: Cache LLVM
+            condition: succeeded()
+
+          - pwsh: |
+              git config --global core.longpaths true
+
+              $directory = Resolve-Path $(LLVM_DIRECTORY)
+
+              # Tell the build where LLVM is installed
+              # This is usually automatic, but we are using a different directory than the default
+              Write-Host "##vso[task.setvariable variable=QIRLIB_LLVM_EXTERNAL_DIR;]$directory"
+
+              # Tell the build what llvm feature version to use
+              $feature = "llvm$(LLVM_VERSION)-0"
+              $variableName = "LLVM_SYS_$(LLVM_VERSION)0_PREFIX"
+              Write-Host "##vso[task.setvariable variable=PYQIR_LLVM_FEATURE_VERSION;]$feature"
+
+              # Tell LLVM-SYS where LLVM is installed
+              Write-Host "##vso[task.setvariable variable=$variableName;]$directory"
+
+              # Tell the qirlib build where to cache the LLVM directory and to not download LLVM
+              Write-Host "##vso[task.setvariable variable=QIRLIB_CACHE_DIR;]$directory"
+              Write-Host "##vso[task.setvariable variable=QIRLIB_DOWNLOAD_LLVM;]false"
+            displayName: Configure Environment
+            condition: succeeded()
+
+          # prerequisites for building
+          - script: |
+              sudo apt-get install -y ninja-build
+            displayName: Install build dependencies
+            condition: and(succeeded(), eq(variables['Agent.OS'], 'Linux'), ne(variables.CACHE_RESTORED, 'true'))
```

Comment:
Is all the caching stuff new too?

Reply:
Caching was used in the sdl pipeline but I hadn't yet brought it over to the release until now.